services/scheduler/scheduler.go

```diff
@@ -0,0 +1,327 @@
+		})
+		if err != nil {
+			s.l.WithField("id", id).Errorf("failed to set next run for new created task")
+		}
+
+		return nil
+	})
+	return scheduledTask, err
+}
+
+// Remove stops task specified by id and removes it from DB and scheduler.
+func (s *Service) Remove(id string) error {
+	s.taskMx.RLock()
+	if cancel, ok := s.tasks[id]; ok {
+		cancel()
+	}
+	s.taskMx.RUnlock()
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		if err := models.RemoveScheduledTask(tx.Querier, id); err != nil {
+			return err
+		}
+
+		_ = s.scheduler.RemoveByTag(id)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reload removes job from scheduler and add it again from DB.
+func (s *Service) Reload(id string) error {
+	dbTask, err := models.FindScheduledTaskByID(s.db.Querier, id)
+	if err != nil {
+		return err
+	}
+
+	if dbTask.Running {
+		return fmt.Errorf("task is running")
+	}
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	task, err := s.convertDBTask(dbTask)
+	if err != nil {
+		return err
+	}
+
+	_ = s.scheduler.RemoveByTag(id)
+
+	// Don't add it to scheduler, if it's disabled.
+	if dbTask.Disabled {
+		return nil
+	}
+
+	j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+	if !dbTask.StartAt.IsZero() {
+		j = j.StartAt(dbTask.StartAt)
+	}
+
+	fn := s.wrapTask(task, dbTask.ID)
+	if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	disabled := false
+	dbTasks, err := models.FindScheduledTasks(s.db.Querier, models.ScheduledTasksFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	tasks := make([]Task, 0, len(dbTasks))
+	for _, dbTask := range dbTasks {
+		task, err := s.convertDBTask(dbTask)
+		if err != nil {
+			return err
+		}
+		tasks = append(tasks, task)
+	}
+
+	s.scheduler.Clear()
+	for i, task := range tasks {
+		dbTask := dbTasks[i]
+		fn := s.wrapTask(task, dbTask.ID)
+		j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+		if !dbTask.StartAt.IsZero() {
+			j = j.StartAt(dbTask.StartAt)
+		}
+		if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (s *Service) wrapTask(task Task, id string) func() {
+	return func() {
+		var err error
+		l := s.l.WithFields(logrus.Fields{
+			"id":       id,
+			"taskType": task.Type(),
+		})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.taskMx.Lock()
+		s.tasks[id] = cancel
+		s.taskMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.taskMx.Lock()
+			delete(s.tasks, id)
+			s.taskMx.Unlock()
+		}()
+
+		t := time.Now()
+		l.Debug("Starting task")
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(true),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running state: %v", err)
+		}
+
+		taskErr := task.Run(ctx)
```

Comment:
task might be outdated(changed) to this moment, so it's better to reload it first.

Reply:
my suggestion to convert DBTask into Task in `wrapTask` and keep using it here.

Reply:
If DB entry is changed, `Reload` should be called.
If I add load from DB to wrap task, then there will be DB select many times unnecessary, even though nothing changed.

What do you think @artemgavrilov @oter.

Reply:
I agree with @BupycHuk

Reply:
I need a few minutes to figure out the problem. 🤔

Reply:
It looks like a task should be added to the scheduler(library) inside a db transaction. For example, if we will have two concurrent `Reload calls` it may look like this:
```
load_from_db_1 ---> load_from_db_2 ---> update_scheduller_2 ---> update_scheduller_1
```

In this situation, we will have the new version in DB, but the old version is the scheduler.
So it looks that this pair `load_from_db, update_scheduler` should be always wrapped into TX, no?

Reply:
|  if it should reload when task is actually run. But from my point of view, I don't think it's necessary.
The current implementation of `Reload` allows already triggered tasks to finish their jobs, am I right?

Reply:
Yes, you can't reload if task is running.

Reply:
Well, it sounds good to me.

Reply:
Oh, what do you mean by `can't`?

Reply:
Right now there is
```
	if dbTask.Running {
		return errors.New("task is running")
	}
```
But probably it's not necessary. If task is running it will finish with old data.

Reply:
With artem we discussed that it will be better have DB update and scheduler internal update inside one transaction. So I rename `Reload` to `Update` and now it also updates DB and then reload it immediately. 

So let's say I have already running task. I'll update it. Running task is still running with old data. But future tasks will be changed.

I don't see point now, to select DB task inside wrap function.